In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('car_price.csv')

In [3]:
df.head()

CarName      carbody drivewheel enginelocation fueltype  \
0        alfa-romero giulia  convertible        rwd          front      gas   
1       alfa-romero stelvio  convertible        rwd          front      gas   
2  alfa-romero Quadrifoglio    hatchback        rwd          front      gas   
3               audi 100 ls        sedan        fwd          front      gas   
4                audi 100ls        sedan        4wd          front      gas   

  aspiration doornumber cylindernumber enginetype fuelsystem  ...  curbweight  \
0        std        two           four       dohc       mpfi  ...        2548   
1        std        two           four       dohc       mpfi  ...        2548   
2        std        two            six       ohcv       mpfi  ...        2823   
3        std       four           four        ohc       mpfi  ...        2337   
4        std       four           five        ohc       mpfi  ...        2824   

   enginesize  boreratio  stroke  compressionratio  horsepower  peakrpm  \
0         130       3.47    2.68               9.0         111     5000   
1         130       3.47    2.68               9.0         111     5000   
2         152       2.68    3.47               9.0         154     5000   
3         109       3.19    3.40              10.0         102     5500   
4         136       3.19    3.40               8.0         115     5500   

   citympg  highwaympg    price  
0       21          27  13495.0  
1       21          27  16500.0  
2       19          26  16500.0  
3       24          30  13950.0  
4       18          22  17450.0  

[5 rows x 25 columns]

In [4]:
df.shape

(210, 25)

In [5]:
df.columns

Index(['CarName', 'carbody', 'drivewheel', 'enginelocation', 'fueltype',
       'aspiration', 'doornumber', 'cylindernumber', 'enginetype',
       'fuelsystem', 'symboling', 'wheelbase', 'carlength', 'carwidth',
       'carheight', 'curbweight', 'enginesize', 'boreratio', 'stroke',
       'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg',
       'price'],
      dtype='object')

In [6]:
df['CarName'].unique()

array(['alfa-romero giulia', 'alfa-romero stelvio',
       'alfa-romero Quadrifoglio', 'audi 100 ls', 'audi 100ls',
       'audi fox', 'audi 5000', 'audi 4000', 'audi 5000s (diesel)',
       'bmw 320i', 'bmw x1', 'bmw x3', 'bmw z4', 'bmw x4', 'bmw x5',
       'chevrolet impala', 'chevrolet monte carlo', 'chevrolet vega 2300',
       'dodge rampage', 'dodge challenger se', 'dodge d200',
       'dodge monaco (sw)', 'dodge colt hardtop', 'dodge colt (sw)',
       'dodge coronet custom', 'dodge dart custom',
       'dodge coronet custom (sw)', 'honda civic', 'honda civic cvcc',
       'honda accord cvcc', 'honda accord lx', 'honda civic 1500 gl',
       'honda accord', 'honda civic 1300', 'honda prelude',
       'honda civic (auto)', 'isuzu MU-X', 'isuzu D-Max ',
       'isuzu D-Max V-Cross', 'jaguar xj', 'jaguar xf', 'jaguar xk',
       'maxda rx3', 'maxda glc deluxe', 'mazda rx2 coupe', 'mazda rx-4',
       'mazda glc deluxe', 'mazda 626', 'mazda glc', 'mazda rx-7 gs',
       'mazda glc 

In [7]:
df['carbody'].unique()

array(['convertible', 'hatchback', 'sedan', 'wagon', 'hardtop'],
      dtype=object)

In [8]:
df['drivewheel'].unique()

array(['rwd', 'fwd', '4wd'], dtype=object)

In [9]:
df['enginelocation'].unique()

array(['front', 'rear'], dtype=object)

In [10]:
df.columns

Index(['CarName', 'carbody', 'drivewheel', 'enginelocation', 'fueltype',
       'aspiration', 'doornumber', 'cylindernumber', 'enginetype',
       'fuelsystem', 'symboling', 'wheelbase', 'carlength', 'carwidth',
       'carheight', 'curbweight', 'enginesize', 'boreratio', 'stroke',
       'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg',
       'price'],
      dtype='object')

In [11]:
df['fueltype'].unique()

array(['gas', 'diesel'], dtype=object)

In [12]:
df.head(1)

CarName      carbody drivewheel enginelocation fueltype  \
0  alfa-romero giulia  convertible        rwd          front      gas   

  aspiration doornumber cylindernumber enginetype fuelsystem  ...  curbweight  \
0        std        two           four       dohc       mpfi  ...        2548   

   enginesize  boreratio  stroke  compressionratio  horsepower  peakrpm  \
0         130       3.47    2.68               9.0         111     5000   

   citympg  highwaympg    price  
0       21          27  13495.0  

[1 rows x 25 columns]

In [13]:
df['aspiration'].unique()

array(['std', 'turbo'], dtype=object)

In [14]:
df['doornumber'].unique()

array(['two', 'four'], dtype=object)

In [15]:
df['cylindernumber'].unique()

array(['four', 'six', 'five', 'three', 'twelve', 'two', 'eight'],
      dtype=object)

In [16]:
df['enginetype'].unique()

array(['dohc', 'ohcv', 'ohc', 'l', 'rotor', 'ohcf', 'dohcv'], dtype=object)

In [17]:
df['fuelsystem'].unique()

array(['mpfi', '2bbl', 'mfi', '1bbl', 'spfi', '4bbl', 'idi', 'spdi'],
      dtype=object)

In [18]:
df.columns

Index(['CarName', 'carbody', 'drivewheel', 'enginelocation', 'fueltype',
       'aspiration', 'doornumber', 'cylindernumber', 'enginetype',
       'fuelsystem', 'symboling', 'wheelbase', 'carlength', 'carwidth',
       'carheight', 'curbweight', 'enginesize', 'boreratio', 'stroke',
       'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg',
       'price'],
      dtype='object')

In [19]:
df.drop(columns=['CarName','carbody','enginelocation','carlength','carwidth','carheight','curbweight','drivewheel'], inplace=True)

In [20]:
df.head()

fueltype aspiration doornumber cylindernumber enginetype fuelsystem  \
0      gas        std        two           four       dohc       mpfi   
1      gas        std        two           four       dohc       mpfi   
2      gas        std        two            six       ohcv       mpfi   
3      gas        std       four           four        ohc       mpfi   
4      gas        std       four           five        ohc       mpfi   

   symboling  wheelbase  enginesize  boreratio  stroke  compressionratio  \
0          3       88.6         130       3.47    2.68               9.0   
1          3       88.6         130       3.47    2.68               9.0   
2          1       94.5         152       2.68    3.47               9.0   
3          2       99.8         109       3.19    3.40              10.0   
4          2       99.4         136       3.19    3.40               8.0   

   horsepower  peakrpm  citympg  highwaympg    price  
0         111     5000       21          27  13495.0  
1         111     5000       21          27  16500.0  
2         154     5000       19          26  16500.0  
3         102     5500       24          30  13950.0  
4         115     5500       18          22  17450.0

In [22]:
df.isna().sum()

fueltype            0
aspiration          0
doornumber          0
cylindernumber      0
enginetype          0
fuelsystem          0
symboling           0
wheelbase           0
enginesize          0
boreratio           0
stroke              0
compressionratio    0
horsepower          0
peakrpm             0
citympg             0
highwaympg          0
price               0
dtype: int64

In [23]:
df.duplicated().sum()

7

In [24]:
df.drop_duplicates(inplace=True)

In [25]:
df.describe()

symboling   wheelbase  enginesize   boreratio      stroke  \
count  203.000000  203.000000  203.000000  203.000000  203.000000   
mean     0.832512   98.806404  127.270936    3.333300    3.255665   
std      1.251347    6.030353   41.685295    0.269794    0.315135   
min     -2.000000   86.600000   61.000000    2.540000    2.070000   
25%      0.000000   94.500000   98.000000    3.150000    3.110000   
50%      1.000000   97.000000  120.000000    3.310000    3.290000   
75%      2.000000  102.400000  143.000000    3.585000    3.410000   
max      3.000000  120.900000  326.000000    3.940000    4.170000   

       compressionratio  horsepower      peakrpm     citympg  highwaympg  \
count        203.000000  203.000000   203.000000  203.000000  203.000000   
mean          10.149852  104.472906  5121.428571   25.162562   30.679803   
std            3.990965   39.575007   477.873206    6.548991    6.882388   
min            7.000000   48.000000  4150.000000   13.000000   16.000000   
25%            8.550000   70.000000  4800.000000   19.000000   25.000000   
50%            9.000000   95.000000  5200.000000   24.000000   30.000000   
75%            9.400000  116.000000  5500.000000   30.000000   34.000000   
max           23.000000  288.000000  6600.000000   49.000000   54.000000   

              price  
count    203.000000  
mean   13343.865355  
std     7999.286862  
min     5118.000000  
25%     7847.000000  
50%    10345.000000  
75%    16509.000000  
max    45400.000000

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 203 entries, 0 to 204
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   fueltype          203 non-null    object 
 1   aspiration        203 non-null    object 
 2   doornumber        203 non-null    object 
 3   cylindernumber    203 non-null    object 
 4   enginetype        203 non-null    object 
 5   fuelsystem        203 non-null    object 
 6   symboling         203 non-null    int64  
 7   wheelbase         203 non-null    float64
 8   enginesize        203 non-null    int64  
 9   boreratio         203 non-null    float64
 10  stroke            203 non-null    float64
 11  compressionratio  203 non-null    float64
 12  horsepower        203 non-null    int64  
 13  peakrpm           203 non-null    int64  
 14  citympg           203 non-null    int64  
 15  highwaympg        203 non-null    int64  
 16  price             203 non-null    float64
dtypes:

In [27]:
import seaborn as sns
import matplotlib.pyplot as plt

In [28]:
df.head(1)

fueltype aspiration doornumber cylindernumber enginetype fuelsystem  \
0      gas        std        two           four       dohc       mpfi   

   symboling  wheelbase  enginesize  boreratio  stroke  compressionratio  \
0          3       88.6         130       3.47    2.68               9.0   

   horsepower  peakrpm  citympg  highwaympg    price  
0         111     5000       21          27  13495.0

In [29]:
df.shape

(203, 17)

In [30]:
df['doornumber'].replace({'two':2, 'four':4}, inplace=True)

In [31]:
df['cylindernumber'].replace({'two':2,'three':3, 'four':4,'five':5, 'six':6, 'eight':8, 'twelve':12}, inplace=True)

In [32]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()


In [33]:
df.columns

Index(['fueltype', 'aspiration', 'doornumber', 'cylindernumber', 'enginetype',
       'fuelsystem', 'symboling', 'wheelbase', 'enginesize', 'boreratio',
       'stroke', 'compressionratio', 'horsepower', 'peakrpm', 'citympg',
       'highwaympg', 'price'],
      dtype='object')

In [34]:
df['fueltype']=le.fit_transform(df['fueltype'])

In [35]:
df['aspiration']=le.fit_transform(df['aspiration'])

In [36]:
df['enginetype']=le.fit_transform(df['enginetype'])

In [37]:
df['fuelsystem']=le.fit_transform(df['fuelsystem'])

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 203 entries, 0 to 204
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   fueltype          203 non-null    int32  
 1   aspiration        203 non-null    int32  
 2   doornumber        203 non-null    int64  
 3   cylindernumber    203 non-null    int64  
 4   enginetype        203 non-null    int32  
 5   fuelsystem        203 non-null    int32  
 6   symboling         203 non-null    int64  
 7   wheelbase         203 non-null    float64
 8   enginesize        203 non-null    int64  
 9   boreratio         203 non-null    float64
 10  stroke            203 non-null    float64
 11  compressionratio  203 non-null    float64
 12  horsepower        203 non-null    int64  
 13  peakrpm           203 non-null    int64  
 14  citympg           203 non-null    int64  
 15  highwaympg        203 non-null    int64  
 16  price             203 non-null    float64
dtypes:

In [39]:
x = df.iloc[:,0:16]
y = df.iloc[:,-1]

In [40]:
x.head()

fueltype  aspiration  doornumber  cylindernumber  enginetype  fuelsystem  \
0         1           0           2               4           0           5   
1         1           0           2               4           0           5   
2         1           0           2               6           5           5   
3         1           0           4               4           3           5   
4         1           0           4               5           3           5   

   symboling  wheelbase  enginesize  boreratio  stroke  compressionratio  \
0          3       88.6         130       3.47    2.68               9.0   
1          3       88.6         130       3.47    2.68               9.0   
2          1       94.5         152       2.68    3.47               9.0   
3          2       99.8         109       3.19    3.40              10.0   
4          2       99.4         136       3.19    3.40               8.0   

   horsepower  peakrpm  citympg  highwaympg  
0         111     5000       21          27  
1         111     5000       21          27  
2         154     5000       19          26  
3         102     5500       24          30  
4         115     5500       18          22

In [41]:
y.head()

0    13495.0
1    16500.0
2    16500.0
3    13950.0
4    17450.0
Name: price, dtype: float64

# Train Test Split

In [42]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x, y, test_size=0.25, random_state=0)


In [43]:
x_train.head()

fueltype  aspiration  doornumber  cylindernumber  enginetype  fuelsystem  \
181         1           0           4               6           0           5   
198         1           1           4               4           3           5   
146         1           0           4               4           4           1   
139         1           0           2               4           4           1   
193         1           0           4               4           3           5   

     symboling  wheelbase  enginesize  boreratio  stroke  compressionratio  \
181         -1      104.5         161       3.27    3.35               9.2   
198         -2      104.3         130       3.62    3.15               7.5   
146          0       97.0         108       3.62    2.64               9.0   
139          2       93.7         108       3.62    2.64               8.7   
193          0      100.4         109       3.19    3.40               9.0   

     horsepower  peakrpm  citympg  highwaympg  
181         156     5200       19          24  
198         162     5100       17          22  
146          82     4800       28          32  
139          73     4400       26          31  
193          88     5500       25          31

In [44]:
x_test.head()

fueltype  aspiration  doornumber  cylindernumber  enginetype  fuelsystem  \
18          1           0           2               3           2           1   
45          1           0           4               4           3           1   
33          1           0           2               4           3           0   
37          1           0           2               4           3           0   
109         1           0           4               4           2           5   

     symboling  wheelbase  enginesize  boreratio  stroke  compressionratio  \
18           2       88.4          61       2.91    3.03               9.5   
45           0       94.5          90       3.03    3.11               9.6   
33           1       93.7          92       2.91    3.41               9.2   
37           0       96.5         110       3.15    3.58               9.0   
109          0      114.2         120       3.46    3.19               8.4   

     horsepower  peakrpm  citympg  highwaympg  
18           48     5100       47          53  
45           70     5400       38          43  
33           76     6000       30          34  
37           86     5800       27          33  
109          97     5000       19          24

In [45]:
y_train.head()

181    15750.0
198    18420.0
146     7463.0
139     7053.0
193    12290.0
Name: price, dtype: float64

# Linear Regression Algorithm

In [46]:
from sklearn.linear_model import LinearRegression
lr_model = LinearRegression()

In [47]:
lr_model = lr_model.fit(x_train,y_train)

In [48]:
y_pred = lr_model.predict(x_test)

In [49]:
y_pred

array([  -73.56166731,  5753.33782411,  7564.75859484,  9453.59572137,
       15721.49485103,  8617.85063816, 14835.31650048, 26258.46529238,
       15294.63084602,  6049.71564695, 10284.61409014, 11067.94590571,
       14159.19963351,  6886.72355865,  6698.12326711, 16476.72144072,
        9652.09943622, 13988.62020646, 19174.29090193, 15086.8841209 ,
        5848.48254038,  9974.74660363, 12453.61757323, 17119.42621198,
       10616.20766571, 10310.89483179,  9887.23095809,  9053.17366911,
        6551.64901629,  6539.24670507, 16526.58838599, 10189.46937387,
       26592.34755925, 37513.27170694, 11313.6882049 ,  6025.35850874,
       29313.33186171,  6539.24670507, 15696.3182499 ,  6551.64901629,
       26258.46529238, 18811.83815944, 10914.61977038,  6513.84334388,
       14083.32043578, 22279.32753098, 20944.69431246, 17585.26945815,
        6050.76186993, 14159.19963351,  6326.85703385])

In [50]:
y_test

18      5151.0
45      8916.5
33      6529.0
37      7895.0
109    12440.0
90      7099.0
5      15250.0
126    32528.0
12     20970.0
155     8778.0
61     10595.0
188     9995.0
168     9639.0
162     9258.0
157     7198.0
7      18920.0
131     9895.0
133    12170.0
75     16503.0
66     18344.0
44      8916.5
148     8013.0
63     10795.0
137    18620.0
86      8189.0
143     9960.0
189    11595.0
183     7975.0
163     8058.0
96      7499.0
204    22625.0
60      8495.0
16     41315.0
74     45400.0
80      9959.0
22      6377.0
71     34184.0
97      7999.0
4      17450.0
164     8238.0
127    34028.0
116    17950.0
55     10945.0
26      7609.0
132    11850.0
179    15998.0
110    13860.0
8      23875.0
89      5499.0
171    11549.0
76      5389.0
Name: price, dtype: float64

In [51]:
 from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.8168991675632997

# R2Score is 0.8168991675632997 for Linear Regression

# Ridge regression algorithm

In [52]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV
 
ridge=Ridge()
parameters={'alpha':[1,5,10,20,30,35,40,45,50,55,100]}
ridge_regressor=GridSearchCV(ridge,parameters,scoring='neg_mean_squared_error',cv=5)
ridge_regressor.fit(x_train,y_train)

print(ridge_regressor.best_params_)
print(ridge_regressor.best_score_)


{'alpha': 100}
-10335096.880974827


In [53]:
ridge_predict =ridge_regressor.predict(x_test) 

In [54]:
r2_score(ridge_predict,y_test)

0.658359337036619

# R2Score is 0.0.658359337036619 of Ridge Regression Algorithm

# Lasso regression Algorithm

In [56]:
from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV
 
lasso=Ridge()
parameters={'alpha':[1,5,10,20,30,35,40,45,50,55,100]}
lasso_regressor=GridSearchCV(lasso,parameters,scoring='neg_mean_squared_error',cv=5)
lasso_regressor.fit(x_train,y_train)

print(lasso_regressor.best_params_)
print(lasso_regressor.best_score_)

{'alpha': 100}
-10335096.880974827


In [58]:
lasso_predict =lasso_regressor.predict(x_test)

In [59]:
r2_score(lasso_predict,y_test)

0.658359337036619

# R2Score is 0.658359337036619 of Lasso Regression Algorithm

In [55]:
import pickle


In [56]:
pickle.dump(lr_model, open('lr_model.pkl', 'wb'))

In [57]:
pickle.dump(df, open('df_car_model.pkl', 'wb'))